rts/motoko-rts/src/gc/incremental/phases/update_increment.rs

```diff
@@ -0,0 +1,125 @@
+use crate::{
+    gc::incremental::{
+        array_slicing::slice_array,
+        partitioned_heap::{Partition, PartitionedHeap, PartitionedHeapIterator},
+        roots::visit_roots,
+        time::BoundedTime,
+        Roots, State,
+    },
+    types::*,
+    visitor::visit_pointer_fields,
+};
+
+pub struct UpdateIncrement<'a> {
+    time: &'a mut BoundedTime,
+    heap: &'a PartitionedHeap,
+    iterator: &'a mut PartitionedHeapIterator,
+    updates_needed: bool,
+}
+
+impl<'a> UpdateIncrement<'a> {
+    pub unsafe fn start_phase(state: &mut State) {
+        debug_assert!(state.iterator_state.is_none());
+        let heap = &mut state.partitioned_heap;
+        state.iterator_state = Some(PartitionedHeapIterator::new(heap));
```

Comment:
```
        state.iterator_state = Some(PartitionedHeapIterator::new(heap));
```
Silly question, but where do these Rust objects get allocated and how do we know the GC won't mess with them? IIRC Rust just allocates into Motoko heap blobs - how is the GC prevented from trashing them?

Reply:
This is a very important question. The static variables are declared in the WASM static data. The Motoko linker allocates this in static space and resizes the heap base accordingly.

Reply:
So it actually reserves space for a large struct and overwrites it on new?

Reply:
Yes. `PartitionedHeapIterator::new()` first creates the struct memory on the call stack and then overwrites the static variable space (in static heap).